File: Plugins/RuStoreCore/Source/RuStoreCore/Private/Java/FeatureAvailabilityListenerWrapper.java
// Copyright Epic Games, Inc. All Rights Reserved.

package com.Plugins.RuStoreCore;

import com.Plugins.RuStoreCore.IRuStoreListener;
import ru.rustore.sdk.core.feature.model.FeatureAvailabilityResult;
import ru.rustore.unitysdk.core.callbacks.FeatureAvailabilityListener;

public class FeatureAvailabilityListenerWrapper implements IRuStoreListener, FeatureAvailabilityListener
{
    private Object mutex = new Object();
    private long cppPointer = 0;

    private native void NativeOnFailure(long pointer, Throwable throwable);
    private native void NativeOnSuccess(long pointer, FeatureAvailabilityResult response);

    public FeatureAvailabilityListenerWrapper(long cppPointer) {
        this.cppPointer = cppPointer;
    }

    @Override
    public void OnFailure(Throwable throwable) {
        synchronized (mutex) {
            if (cppPointer != 0) {
                NativeOnFailure(cppPointer, throwable);
            }
        }
    }

    @Override
    public void OnSuccess(FeatureAvailabilityResult response) {
        synchronized (mutex) {
            if (cppPointer != 0) {
                NativeOnSuccess(cppPointer, response);
            }
        }
    }

    public void DisposeCppPointer() {
        synchronized (mutex) {
            cppPointer = 0;
        }
    }
}

File: unreal_plugin_libraries/RuStoreUnityCore/src/main/java/ru/rustore/unitysdk/core/wrappers/FeatureAvailabilityListenerWrapper.kt
package ru.rustore.unitysdk.core.wrappers

import ru.rustore.sdk.core.feature.model.FeatureAvailabilityResult
import ru.rustore.unitysdk.core.IRuStoreListener
import ru.rustore.unitysdk.core.callbacks.FeatureAvailabilityListener

class FeatureAvailabilityListenerWrapper(cppPointer: Long) : IRuStoreListener, FeatureAvailabilityListener {
    private val mutex = Any()
    private var cppPointer: Long = 0
    private external fun NativeOnFailure(pointer: Long, throwable: Throwable)
    private external fun NativeOnSuccess(pointer: Long, response: FeatureAvailabilityResult)

    init {
        this.cppPointer = cppPointer
    }

    override fun OnFailure(throwable: Throwable) {
        synchronized(mutex) {
            if (cppPointer != 0L) {
                NativeOnFailure(cppPointer, throwable)
            }
        }
    }

    override fun OnSuccess(response: FeatureAvailabilityResult) {
        synchronized(mutex) {
            if (cppPointer != 0L) {
                NativeOnSuccess(cppPointer, response)
            }
        }
    }

    override fun DisposeCppPointer() {
        synchronized(mutex) { cppPointer = 0 }
    }
}
